clone/internal/clone/clone.go

```diff
@@ -0,0 +1,193 @@
+// Copyright 2021 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// clone contains the core engine for quickly downloading leaves and adding
+// them to the database.
+package clone
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/google/trillian-examples/clone/internal/download"
+	"github.com/google/trillian-examples/clone/logdb"
+)
+
+// Cloner is configured with the desired performance parameters, and then performs
+// work when Clone is called.
+type Cloner struct {
+	workers        uint
+	fetchBatchSize uint
+	writeBatchSize uint
+
+	db *logdb.Database
+}
+
+// NewCloner creates a new Cloner with the given performance parameters and database.
+func NewCloner(wokers, fetchBatchSize, writeBatchSize uint, db *logdb.Database) Cloner {
```

Comment:
How about renaming this package `cloner` and calling this `New`?
(packages often seem to be nouns too)

Reply:
Done.